File: nonce_values.py
import hashlib
import secrets
import statistics

def calculate_sha256(message):
    """
    Calcule le haché SHA-256 d'un message.
    """
    return hashlib.sha256(message.encode()).hexdigest()

def calculate_sha256_with_nonce(message, nonce):
    """
    Calcule le haché SHA-256 d'un message avec un nonce donné.
    """
    message_with_nonce = f"{message}{nonce}"
    return hashlib.sha256(message_with_nonce.encode()).hexdigest()

def find_nonce(message, leading_zeros):
    """
    Recherche un nonce produisant un haché avec les N bits de poids fort égaux à 0.
    """
    nonce_length = 32  # Longueur du nonce en bytes
    while True:
        nonce = secrets.token_bytes(nonce_length)  # Générer un nonce aléatoire
        hash_result = calculate_sha256_with_nonce(message, nonce)
        if hash_result.startswith('0' * leading_zeros):
            return int.from_bytes(nonce, 'big')  # Convertir le nonce hexadécimal en entier pour N=4 et N=12

def execute_find_nonce(message, leading_zeros, attempts=10):
    """
    Exécute la recherche de nonce plusieurs fois et établit des statistiques.
    """
    nonce_values = []
    for _ in range(attempts):
        nonce = find_nonce(message, leading_zeros)
        if nonce is not None:
            nonce_values.append(nonce)
    return nonce_values

# Message à utiliser
message = "Hello, world!"

# Dictionnaire pour stocker les nonces pour chaque valeur de N
nonce_statistics = {}

# Exécution pour N=4, 8 et 12
for leading_zeros in [4, 8, 12]:
    nonce_values = execute_find_nonce(message, leading_zeros, attempts=10)
    nonce_statistics[leading_zeros] = nonce_values
    print(f"N={leading_zeros}:")
    print("Nonce values:", nonce_values)
    print("Number of nonces:", len(nonce_values))
    print("Average nonce:", statistics.mean(nonce_values))
    print("Standard deviation of nonce:", statistics.stdev(nonce_values))
    print("")

# Affichage des statistiques globales
print("Overall statistics:")
for leading_zeros, nonce_values in nonce_statistics.items():
    print(f"N={leading_zeros}:")
    print("Number of nonces:", len(nonce_values))
    print("Average nonce:", statistics.mean(nonce_values))
    print("Standard deviation of nonce:", statistics.stdev(nonce_values))
    print("")

File: nonce_values.go
package main

import (
	"crypto/rand"
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"math/big"
	"math/bits"
)

// calculateSHA256WithNonce calcule le haché SHA-256 d'un message avec un nonce donné.
func calculateSHA256WithNonce(message string, nonce *big.Int) string {
	messageWithNonce := fmt.Sprintf("%s%d", message, nonce)
	hash := sha256.Sum256([]byte(messageWithNonce))
	return hex.EncodeToString(hash[:])
}

// findNonce recherche un nonce produisant un haché avec les N bits de poids fort égaux à 0.
func findNonce(message string, leadingZeros int) *big.Int {
	nonce := new(big.Int)
	maxNonce := new(big.Int).Lsh(big.NewInt(1), 64) // Max nonce value (2^64)
	for {
		nonce, _ = rand.Int(rand.Reader, maxNonce)
		hashResult := calculateSHA256WithNonce(message, nonce)
		if leadingZeros == 0 || countLeadingZeros(hashResult) >= leadingZeros {
			return nonce
		}
	}
}

// countLeadingZeros compte le nombre de zéros de poids fort dans une chaîne hexadécimale.
func countLeadingZeros(hexString string) int {
	count := 0
	for _, c := range hexString {
		if c != '0' {
			break
		}
		count += 4 // 1 caractère hexadécimal = 4 bits
	}
	return count
}

// calculateStandardDeviation calcule l'écart-type d'un ensemble de valeurs.
func calculateStandardDeviation(values []*big.Int) float64 {
	if len(values) == 0 {
		return 0.0
	}
	var sum int64
	for _, v := range values {
		sum += v.Int64()
	}
	mean := float64(sum) / float64(len(values))
	var squaredDiffSum float64
	for _, v := range values {
		diff := float64(v.Int64()) - mean
		squaredDiffSum += diff * diff
	}
	variance := squaredDiffSum / float64(len(values))
	return float64(bits.Len64(uint64(len(values)))) * (variance / float64(len(values)))
}

func main() {
	// Tableau pour stocker les messages pour chaque valeur de N
	messages := make(map[int][]string)
	nonces := make(map[int][]*big.Int)

	// Générer 10 messages aléatoires pour chaque valeur de N
	for _, leadingZeros := range []int{4, 8, 12} {
		var messageList []string
		var nonceList []*big.Int
		for i := 0; i < 10; i++ {
			message := generateRandomMessage()
			messageList = append(messageList, message)
			nonce := findNonce(message, leadingZeros)
			nonceList = append(nonceList, nonce)
		}
		messages[leadingZeros] = messageList
		nonces[leadingZeros] = nonceList
	}

	// Afficher les messages et les nonces générés
	for leadingZeros, messageList := range messages {
		fmt.Printf("N=%d:\n", leadingZeros)
		for i, message := range messageList {
			fmt.Printf("Message %d: %s\n", i+1, message)
			fmt.Printf("Nonce: %d\n", nonces[leadingZeros][i])
			fmt.Println()
		}
	}

	// Afficher les statistiques pour chaque valeur de N
	for leadingZeros, nonceList := range nonces {
		fmt.Printf("N=%d:\n", leadingZeros)
		fmt.Printf("Number of nonces: %d\n", len(nonceList))
		if len(nonceList) > 0 {
			var sum int64
			for _, nonce := range nonceList {
				sum += nonce.Int64()
			}
			average := float64(sum) / float64(len(nonceList))
			fmt.Printf("Average nonce: %.2f\n", average)
			standardDeviation := calculateStandardDeviation(nonceList)
			fmt.Printf("Standard deviation of nonce: %.2f\n", standardDeviation)
		}
		fmt.Println()
	}
}

// generateRandomMessage génère un message aléatoire.
func generateRandomMessage() string {
	message := make([]byte, 16)
	rand.Read(message)
	return hex.EncodeToString(message)
}
